<img align="center" src="https://iili.io/3wI8gI.png" style="height:90px" style="width:30px"/>

# Preparing Mark Twain's `The Adventures of Tom Sawyers` for text analysis
### By using the built-in functions `Map, Reduce, Filter`

In [24]:
from functools import reduce 
import numpy as np
import pandas as pd
import os

# Challenge 1 - Mapping

#### We will use the map function to clean up words in a book.

In the following cell, we will read a text file containing the book The Prophet by Khalil Gibran.

In [9]:
os.getcwd()


'/Users/hugocarvalho/Documents/nod_coding_bootcamp/lab/week_2'

In [2]:
# Run this cell to read txt file
# You might have to change the location variable depending on where your storing the .txt file

location = '/Users/hugocarvalho/Documents/nod_coding_bootcamp/lab/week_2/data/the_adventures_of_tom_sawyers.txt'
with open(location, 'r', encoding="utf8") as f:
    tom_sawyer = f.read().split(' ')


#### The first 838 words contain information that is not part of the actual book, so let's remove these words.
Do this by removing from `tom_sawyers` elements 0 through 838 of the list, or simply keep the elements starting  from index 838 to the last element.

In [3]:
del tom_sawyer[:838]


In [4]:
tom_sawyer

['1876.\n\n\n\n\nCHAPTER',
 'I\n\n“TOM!”\n\nNo',
 'answer.\n\n“TOM!”\n\nNo',
 "answer.\n\n“What's",
 'gone',
 'with',
 'that',
 'boy,',
 '\xa0I',
 'wonder?',
 'You',
 'TOM!”\n\nNo',
 'answer.\n\nThe',
 'old',
 'lady',
 'pulled',
 'her',
 'spectacles',
 'down',
 'and',
 'looked',
 'over',
 'them',
 'about',
 'the\nroom;',
 'then',
 'she',
 'put',
 'them',
 'up',
 'and',
 'looked',
 'out',
 'under',
 'them.',
 'She',
 'seldom',
 'or\nnever',
 'looked',
 '_through_',
 'them',
 'for',
 'so',
 'small',
 'a',
 'thing',
 'as',
 'a',
 'boy;',
 'they',
 'were\nher',
 'state',
 'pair,',
 'the',
 'pride',
 'of',
 'her',
 'heart,',
 'and',
 'were',
 'built',
 'for',
 '“style,”',
 'not\nservice--she',
 'could',
 'have',
 'seen',
 'through',
 'a',
 'pair',
 'of',
 'stove-lids',
 'just',
 'as',
 'well.\nShe',
 'looked',
 'perplexed',
 'for',
 'a',
 'moment,',
 'and',
 'then',
 'said,',
 'not',
 'fiercely,',
 'but\nstill',
 'loud',
 'enough',
 'for',
 'the',
 'furniture',
 'to',
 'hear:\n\n“Well,',
 '

If you look through the words, you will find that many words have a underscores surounding them. 

In [5]:
# Check for underscores in this cell:
[i for i in tom_sawyer if "_" in i]

['_through_',
 '_is_',
 '_got_',
 '_This_',
 '_make_',
 'will.”\n\n“Much--much--_much_.',
 "_don't_",
 '_do_',
 '_say_',
 '_will_,',
 '_some_,',
 "_don't_",
 '_course_',
 '_will_.”\n\n“Well',
 '_do_',
 '_saying_',
 '_do_',
 "_ain't_",
 'lie.”\n\n“_Your_',
 '_said_',
 '_will_',
 "_she'd_",
 '_She_',
 'would.”\n\n“_She_!',
 '_work_,',
 '_lively_',
 '“_Hi-Yi!',
 "You're_",
 "_work_--wouldn't",
 '_that_',
 '_like_',
 '_me_',
 '_she_',
 'let\n_you_,',
 '_all_',
 '_obliged_',
 '_is_',
 '_me_',
 '_she_',
 'they--they--”\n\n“_Theirs_--”\n\n“For',
 '_theirs_.',
 'is!”\n\n“_Shall_!”\n\n“Oh,',
 '_shall_!',
 '_what_?',
 '_is_',
 '_did_',
 'were--”\n\n“_David',
 'Goliah!_”\n\nLet',
 '_beds_',
 '_blood_-y',
 "_Don't_",
 "_don't!_",
 '_seemed_',
 '_is_',
 '_him_.',
 "_wouldn't_",
 '_any_',
 '_know_',
 'midnight?--and\n_then_',
 '_the',
 'place_',
 'said:\n\n“_I',
 'Finn!_”\n\nThe',
 '_any_body.',
 '_you_',
 '_will_',
 '“_I',
 'you_.”\n\n“Oh,',
 'too--_live_',
 '_you_?”\n\n“Ye--yes--but',
 '_now_.',
 

#### The next step is to create a function that will remove underscores. 

The underscores are there to make the text italic. However, since we just want the plain text, we are going to replace the underscores with an empty string. Write your function below.

In [6]:
def underscore(x):
    return x.replace("_","")
    

    '''
    Input: A string
    Output: The string with underscores removed
    
    Example:
    Input: '_under_'
    Output: 'under'
    '''
    pass

In [7]:
test_word = "___hugo"
test_word = underscore(test_word)
test_word

'hugo'

Now that we have our function, use the `map()` function to apply this function to our book, The Prophet. Return the resulting list to a new list called `sawyer_underscore`.

In [8]:
sawyer_underscore = list(map(lambda x: underscore(x) ,tom_sawyer))

In [9]:
sawyer_underscore

['1876.\n\n\n\n\nCHAPTER',
 'I\n\n“TOM!”\n\nNo',
 'answer.\n\n“TOM!”\n\nNo',
 "answer.\n\n“What's",
 'gone',
 'with',
 'that',
 'boy,',
 '\xa0I',
 'wonder?',
 'You',
 'TOM!”\n\nNo',
 'answer.\n\nThe',
 'old',
 'lady',
 'pulled',
 'her',
 'spectacles',
 'down',
 'and',
 'looked',
 'over',
 'them',
 'about',
 'the\nroom;',
 'then',
 'she',
 'put',
 'them',
 'up',
 'and',
 'looked',
 'out',
 'under',
 'them.',
 'She',
 'seldom',
 'or\nnever',
 'looked',
 'through',
 'them',
 'for',
 'so',
 'small',
 'a',
 'thing',
 'as',
 'a',
 'boy;',
 'they',
 'were\nher',
 'state',
 'pair,',
 'the',
 'pride',
 'of',
 'her',
 'heart,',
 'and',
 'were',
 'built',
 'for',
 '“style,”',
 'not\nservice--she',
 'could',
 'have',
 'seen',
 'through',
 'a',
 'pair',
 'of',
 'stove-lids',
 'just',
 'as',
 'well.\nShe',
 'looked',
 'perplexed',
 'for',
 'a',
 'moment,',
 'and',
 'then',
 'said,',
 'not',
 'fiercely,',
 'but\nstill',
 'loud',
 'enough',
 'for',
 'the',
 'furniture',
 'to',
 'hear:\n\n“Well,',
 'I'

Another thing you may have noticed is that some words contain a line break. Let's write a function to split those words. Our function will return the string split on the character `\n`. Write your function in the cell below.

Have you noticed that some words contain a line break? Yeah, we want to get rid of those as well. You can do this by writing a function that splits on line breaks

In [10]:
def line_breaks(x):
    return x.replace("\n","")
    '''
    Input: A string
    Output: A list of strings split on the line break (\n) character
        
    Example:
    Input: 'answer.\n\nThe'
    Output: ['answer', 'beloved']
    '''
    pass


Apply the `line_breaks` function to the `sawyer_underscore` list. Name the new list `sawyer_line`.

In [11]:
sawyer_line = list(map(lambda x: line_breaks(x) ,tom_sawyer))
sawyer_line

['1876.CHAPTER',
 'I“TOM!”No',
 'answer.“TOM!”No',
 "answer.“What's",
 'gone',
 'with',
 'that',
 'boy,',
 '\xa0I',
 'wonder?',
 'You',
 'TOM!”No',
 'answer.The',
 'old',
 'lady',
 'pulled',
 'her',
 'spectacles',
 'down',
 'and',
 'looked',
 'over',
 'them',
 'about',
 'theroom;',
 'then',
 'she',
 'put',
 'them',
 'up',
 'and',
 'looked',
 'out',
 'under',
 'them.',
 'She',
 'seldom',
 'ornever',
 'looked',
 '_through_',
 'them',
 'for',
 'so',
 'small',
 'a',
 'thing',
 'as',
 'a',
 'boy;',
 'they',
 'wereher',
 'state',
 'pair,',
 'the',
 'pride',
 'of',
 'her',
 'heart,',
 'and',
 'were',
 'built',
 'for',
 '“style,”',
 'notservice--she',
 'could',
 'have',
 'seen',
 'through',
 'a',
 'pair',
 'of',
 'stove-lids',
 'just',
 'as',
 'well.She',
 'looked',
 'perplexed',
 'for',
 'a',
 'moment,',
 'and',
 'then',
 'said,',
 'not',
 'fiercely,',
 'butstill',
 'loud',
 'enough',
 'for',
 'the',
 'furniture',
 'to',
 'hear:“Well,',
 'I',
 'lay',
 'if',
 'I',
 'get',
 'hold',
 'of',
 'you

If you look at the elements of `sawyer_line`, you will see that the function returned lists and not strings. Our list is now a list of lists. Flatten the list using list comprehension. Assign this new list to `sawyer_flat`.

In [12]:
type(sawyer_line)

list

In [13]:
type(sawyer_line[1])

str

The first element of `sawyer_line` is `'1876'` and is actually not a part of the book. Remove the first element.

In [14]:
del sawyer_line[:1]

In [15]:
sawyer_line

['I“TOM!”No',
 'answer.“TOM!”No',
 "answer.“What's",
 'gone',
 'with',
 'that',
 'boy,',
 '\xa0I',
 'wonder?',
 'You',
 'TOM!”No',
 'answer.The',
 'old',
 'lady',
 'pulled',
 'her',
 'spectacles',
 'down',
 'and',
 'looked',
 'over',
 'them',
 'about',
 'theroom;',
 'then',
 'she',
 'put',
 'them',
 'up',
 'and',
 'looked',
 'out',
 'under',
 'them.',
 'She',
 'seldom',
 'ornever',
 'looked',
 '_through_',
 'them',
 'for',
 'so',
 'small',
 'a',
 'thing',
 'as',
 'a',
 'boy;',
 'they',
 'wereher',
 'state',
 'pair,',
 'the',
 'pride',
 'of',
 'her',
 'heart,',
 'and',
 'were',
 'built',
 'for',
 '“style,”',
 'notservice--she',
 'could',
 'have',
 'seen',
 'through',
 'a',
 'pair',
 'of',
 'stove-lids',
 'just',
 'as',
 'well.She',
 'looked',
 'perplexed',
 'for',
 'a',
 'moment,',
 'and',
 'then',
 'said,',
 'not',
 'fiercely,',
 'butstill',
 'loud',
 'enough',
 'for',
 'the',
 'furniture',
 'to',
 'hear:“Well,',
 'I',
 'lay',
 'if',
 'I',
 'get',
 'hold',
 'of',
 'you',
 "I'll--”She",

# Challenge 2 - Filtering

When printing out a few words from the book, we see that there are words that we may not want to keep if we choose to analyze the corpus of text. Below is a list of words that we would like to get rid of. Create a function that will return false if it contains a word from the list of words specified and true otherwise.

In [16]:
word_list = ['and', 'the', "a", "an"]

In [18]:
def word_filter(x):

    if x in word_list:
        return False

    return True
    '''

    Input: A string
    Output: True if the word is not in the specified list 
    and False if the word is in the list
    '''
    
    # your code here
    

Use the `filter()` function to filter out the words speficied in the `word_filter()` function. Store the filtered list in the variable `sawyer_filter`.

In [19]:
sawyer_filter = list(filter(word_filter,sawyer_line))


In [20]:
sawyer_filter

['I“TOM!”No',
 'answer.“TOM!”No',
 "answer.“What's",
 'gone',
 'with',
 'that',
 'boy,',
 '\xa0I',
 'wonder?',
 'You',
 'TOM!”No',
 'answer.The',
 'old',
 'lady',
 'pulled',
 'her',
 'spectacles',
 'down',
 'looked',
 'over',
 'them',
 'about',
 'theroom;',
 'then',
 'she',
 'put',
 'them',
 'up',
 'looked',
 'out',
 'under',
 'them.',
 'She',
 'seldom',
 'ornever',
 'looked',
 '_through_',
 'them',
 'for',
 'so',
 'small',
 'thing',
 'as',
 'boy;',
 'they',
 'wereher',
 'state',
 'pair,',
 'pride',
 'of',
 'her',
 'heart,',
 'were',
 'built',
 'for',
 '“style,”',
 'notservice--she',
 'could',
 'have',
 'seen',
 'through',
 'pair',
 'of',
 'stove-lids',
 'just',
 'as',
 'well.She',
 'looked',
 'perplexed',
 'for',
 'moment,',
 'then',
 'said,',
 'not',
 'fiercely,',
 'butstill',
 'loud',
 'enough',
 'for',
 'furniture',
 'to',
 'hear:“Well,',
 'I',
 'lay',
 'if',
 'I',
 'get',
 'hold',
 'of',
 'you',
 "I'll--”She",
 'did',
 'not',
 'finish,',
 'for',
 'by',
 'this',
 'time',
 'she',
 '

# Bonus Challenge

Rewrite the `word_filter` function above to not be case sensitive.

In [21]:
def word_filter(x):

    if x.lower() in word_list:
        return False

    return True

# Challenge 3 - Reducing

#### Now that we have significantly cleaned up our text corpus, let's use the `reduce()` function to put the words back together into one long string separated by spaces. 

We will start by writing a function that takes two strings and concatenates them together with a space between the two strings.

In [22]:
def concat_space(a, b):
    return a + " " + b
    

Use the function above to reduce the text corpus in the list `sawyer_filter` into a single string. Assign this new string to the variable `sawyer_string`.

In [25]:
sawyer_string = reduce(concat_space,sawyer_filter)
sawyer_string

"I“TOM!”No answer.“TOM!”No answer.“What's gone with that boy, \xa0I wonder? You TOM!”No answer.The old lady pulled her spectacles down looked over them about theroom; then she put them up looked out under them. She seldom ornever looked _through_ them for so small thing as boy; they wereher state pair, pride of her heart, were built for “style,” notservice--she could have seen through pair of stove-lids just as well.She looked perplexed for moment, then said, not fiercely, butstill loud enough for furniture to hear:“Well, I lay if I get hold of you I'll--”She did not finish, for by this time she was bending down punchingunder bed with broom, so she needed breath to punctuate thepunches with. She resurrected nothing but cat.“I never did see beat of that boy!”She went to open door stood in it looked out among thetomato vines “jimpson” weeds that constituted garden. No Tom. Soshe lifted up her voice at angle calculated for distance shouted:“Y-o-u-u TOM!”There was slight noise behind her s

In [ ]:
"End of Project Gutenberg"

#### As we can se when having reduced the text file to a string, we can see that the end of the string contains information that isn't part of the book. 

Turn `sawyer filter` into a list and slice it in way that makes the information that does not belong to the book itself go away.

In [26]:
sawyer_test = list(sawyer_string)
sawyer_test

['I',
 '“',
 'T',
 'O',
 'M',
 '!',
 '”',
 'N',
 'o',
 ' ',
 'a',
 'n',
 's',
 'w',
 'e',
 'r',
 '.',
 '“',
 'T',
 'O',
 'M',
 '!',
 '”',
 'N',
 'o',
 ' ',
 'a',
 'n',
 's',
 'w',
 'e',
 'r',
 '.',
 '“',
 'W',
 'h',
 'a',
 't',
 "'",
 's',
 ' ',
 'g',
 'o',
 'n',
 'e',
 ' ',
 'w',
 'i',
 't',
 'h',
 ' ',
 't',
 'h',
 'a',
 't',
 ' ',
 'b',
 'o',
 'y',
 ',',
 ' ',
 '\xa0',
 'I',
 ' ',
 'w',
 'o',
 'n',
 'd',
 'e',
 'r',
 '?',
 ' ',
 'Y',
 'o',
 'u',
 ' ',
 'T',
 'O',
 'M',
 '!',
 '”',
 'N',
 'o',
 ' ',
 'a',
 'n',
 's',
 'w',
 'e',
 'r',
 '.',
 'T',
 'h',
 'e',
 ' ',
 'o',
 'l',
 'd',
 ' ',
 'l',
 'a',
 'd',
 'y',
 ' ',
 'p',
 'u',
 'l',
 'l',
 'e',
 'd',
 ' ',
 'h',
 'e',
 'r',
 ' ',
 's',
 'p',
 'e',
 'c',
 't',
 'a',
 'c',
 'l',
 'e',
 's',
 ' ',
 'd',
 'o',
 'w',
 'n',
 ' ',
 'l',
 'o',
 'o',
 'k',
 'e',
 'd',
 ' ',
 'o',
 'v',
 'e',
 'r',
 ' ',
 't',
 'h',
 'e',
 'm',
 ' ',
 'a',
 'b',
 'o',
 'u',
 't',
 ' ',
 't',
 'h',
 'e',
 'r',
 'o',
 'o',
 'm',
 ';',
 ' ',
 't',
 'h',
 'e',
 

In [27]:
end_phrase = "End of Project Gutenberg"
end_phrase_list = list(end_phrase)
end_phrase_list

['E',
 'n',
 'd',
 ' ',
 'o',
 'f',
 ' ',
 'P',
 'r',
 'o',
 'j',
 'e',
 'c',
 't',
 ' ',
 'G',
 'u',
 't',
 'e',
 'n',
 'b',
 'e',
 'r',
 'g']

In [33]:

def index_of_last_part(string_text, end_string):
    for book_index, letter_in_book in enumerate(string_text):
        for list_index, letter_in_end in enumerate(end_string):
            if letter_in_end != string_text[book_index + list_index]:
                break
            elif list_index == len(end_string) - 1:
                return book_index

index_of_last_part(sawyer_string,end_phrase)
sawyer_string[:index_of_last_part(sawyer_string,end_phrase)]




"I“TOM!”No answer.“TOM!”No answer.“What's gone with that boy, \xa0I wonder? You TOM!”No answer.The old lady pulled her spectacles down looked over them about theroom; then she put them up looked out under them. She seldom ornever looked _through_ them for so small thing as boy; they wereher state pair, pride of her heart, were built for “style,” notservice--she could have seen through pair of stove-lids just as well.She looked perplexed for moment, then said, not fiercely, butstill loud enough for furniture to hear:“Well, I lay if I get hold of you I'll--”She did not finish, for by this time she was bending down punchingunder bed with broom, so she needed breath to punctuate thepunches with. She resurrected nothing but cat.“I never did see beat of that boy!”She went to open door stood in it looked out among thetomato vines “jimpson” weeds that constituted garden. No Tom. Soshe lifted up her voice at angle calculated for distance shouted:“Y-o-u-u TOM!”There was slight noise behind her s

In [32]:
def index_of_last_part(string_text, end_string):
    for book_index, letter_in_book in enumerate(string_text):
        for list_index, letter_in_end in enumerate(end_string):
            if letter_in_end != string_text[book_index + list_index]:
                break
            elif list_index == len(end_string) - 1:
                return book_index

index_of_last_part(sawyer_string,end_phrase)

sawyer_string[:index_of_last_part(sawyer_string,end_phrase)]


"I“TOM!”No answer.“TOM!”No answer.“What's gone with that boy, \xa0I wonder? You TOM!”No answer.The old lady pulled her spectacles down looked over them about theroom; then she put them up looked out under them. She seldom ornever looked _through_ them for so small thing as boy; they wereher state pair, pride of her heart, were built for “style,” notservice--she could have seen through pair of stove-lids just as well.She looked perplexed for moment, then said, not fiercely, butstill loud enough for furniture to hear:“Well, I lay if I get hold of you I'll--”She did not finish, for by this time she was bending down punchingunder bed with broom, so she needed breath to punctuate thepunches with. She resurrected nothing but cat.“I never did see beat of that boy!”She went to open door stood in it looked out among thetomato vines “jimpson” weeds that constituted garden. No Tom. Soshe lifted up her voice at angle calculated for distance shouted:“Y-o-u-u TOM!”There was slight noise behind her s

## Done, NOICE!